File: LeetCode/DataStructure_II/Day14/1249_minimum_remove_to_make_valid_parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        letter_list = list(s)
        
        score = 0
        stack = []
        removed = []
        for i in range(len(letter_list)-1, -1, -1):
            letter = letter_list[i]
            if letter == ')':
                score += 1
                stack.append(i)
            elif letter == '(':
                score -= 1
                if score < 0:
                    removed.append(i)
                    score += 1
                if stack:
                    stack.pop()
                    
        remove_index = set(stack + removed)
        letter_list = [letter for idx, letter in enumerate(letter_list) if idx not in remove_index]
        return ''.join(letter_list)
                    

File: LeetCode/Problems/1249_minimum_remove_to_make_valid_parentheses.js
/**
 * @param {string} s
 * @return {string}
 */
 var minRemoveToMakeValid = function(s) {
    let stack = []
    let invalid = [];
    let result = [];
    s.split('').forEach((letter, idx) => {
        if(letter === '(') {
            stack.push(idx)    
        }
        else if(letter === ')') {
            if(stack.length === 0){
                invalid.push(idx)
            }
            stack?.pop()
        }
    })
    console.log(invalid, stack)
    return s.split('').filter((item, idx) => {
        return !(invalid.includes(idx) || stack.includes(idx))
    }).join('')
};